src/xenia/gpu/gl4/gl4_shader.cc

```diff
@@ -120,81 +127,135 @@ bool GL4Shader::PrepareVertexArrayObject() {
   return true;
 }
 
-bool GL4Shader::CompileProgram() {
+bool GL4Shader::CompileShader() {
   assert_zero(program_);
 
-  // Give source to GL.
+  shader_ =
+      glCreateShader(shader_type_ == ShaderType::kVertex ? GL_VERTEX_SHADER
+                                                         : GL_FRAGMENT_SHADER);
+  if (!shader_) {
+    XELOGE("OpenGL could not create a shader object!");
+    return false;
+  }
+
   auto source_str = GetTranslatedBinaryString();
   auto source_str_ptr = source_str.c_str();
-  program_ = glCreateShaderProgramv(shader_type_ == ShaderType::kVertex
-                                        ? GL_VERTEX_SHADER
-                                        : GL_FRAGMENT_SHADER,
-                                    1, &source_str_ptr);
+  GLint source_length = GLint(source_str.length());
+  glShaderSource(shader_, 1, &source_str_ptr, &source_length);
+  glCompileShader(shader_);
+
+  GLint status = 0;
+  glGetShaderiv(shader_, GL_COMPILE_STATUS, &status);
+
+  return status == GL_TRUE;
+}
+
+bool GL4Shader::LinkProgram() {
+  program_ = glCreateProgram();
   if (!program_) {
-    XELOGE("Unable to create shader program");
+    XELOGE("OpenGL could not create a shader program!");
     return false;
   }
 
-  // Output info log.
-  // log_length includes the null character.
-  GLint log_length = 0;
-  glGetProgramiv(program_, GL_INFO_LOG_LENGTH, &log_length);
-  std::string info_log;
-  if (log_length > 0) {
-    info_log.resize(log_length - 1);
-    glGetProgramInfoLog(program_, log_length, &log_length, &info_log[0]);
-  }
+  glAttachShader(program_, shader_);
 
-  if (!info_log.empty()) {
-    XELOGD("Shader log: %s", info_log.c_str());
+  // Enable TFB
+  if (shader_type_ == ShaderType::kVertex) {
+    const GLchar* feedbackVaryings = "gl_Position";
+    glTransformFeedbackVaryings(program_, 1, &feedbackVaryings,
+                                GL_SEPARATE_ATTRIBS);
   }
 
-  // Get error log, if we failed to link.
+  glProgramParameteri(program_, GL_PROGRAM_SEPARABLE, GL_TRUE);
+  glLinkProgram(program_);
+
   GLint link_status = 0;
   glGetProgramiv(program_, GL_LINK_STATUS, &link_status);
   if (!link_status) {
-    host_error_log_ = std::move(info_log);
     assert_always("Unable to link generated shader");
     return false;
   }
 
+  return true;
+}
+
+std::string GL4Shader::GetShaderInfoLog() {
+  if (!shader_) {
+    return "GL4Shader::GetShaderInfoLog(): Program is NULL";
+  }
+
+  std::string log;
+  GLint log_length = 0;
+  glGetShaderiv(shader_, GL_INFO_LOG_LENGTH, &log_length);
+  if (log_length > 0) {
+    log.resize(log_length - 1);
+    glGetShaderInfoLog(shader_, log_length, &log_length, &log[0]);
```

Comment:
Whenever I see code that gets logs from gl functions, they tend to store the log in a vector<char> instead of a std::string. i guess that's to avoid casting like DrChat said